### ISSUE TITLE
make fails on Debian Stretch

### ISSUE THREAD

**Issue by h3r0n** [Ultimaker/cura-build#105]
could you help me?

```
~/cura-build/build$ make
[  0%] Built target Python
[  6%] Built target Sip
[  6%] Performing build step for 'Qt'
ssl/qsslcertificate_openssl.cpp: In function ‘uint qHash(const QSslCertificate&, uint)’:
ssl/qsslcertificate_openssl.cpp:63:30: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
         return qHashBits(x509->sha1_hash, SHA_DIGEST_LENGTH, seed);
                              ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
ssl/qsslcertificate_openssl.cpp: In member function ‘QByteArray QSslCertificate::version() const’:
ssl/qsslcertificate_openssl.cpp:87:68: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
             QByteArray::number(qlonglong(q_ASN1_INTEGER_get(d->x509->cert_info->version)) + 1);
                                                                    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
ssl/qsslcertificate_openssl.cpp: In member function ‘QByteArray QSslCertificate::serialNumber() const’:
ssl/qsslcertificate_openssl.cpp:96:45: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
         ASN1_INTEGER *serialNumber = d->x509->cert_info->serialNumber;
                                             ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
ssl/qsslcertificate_openssl.cpp: In member function ‘QSslKey QSslCertificate::publicKey() const’:
ssl/qsslcertificate_openssl.cpp:232:32: error: invalid use of incomplete type ‘X509 {aka struct x509_st}’
     X509_PUBKEY *xkey = d->x509->cert_info->key;
                                ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:119:16: note: forward declaration of ‘X509 {aka struct x509_st}’
 typedef struct x509_st X509;
                ^~~~~~~
ssl/qsslcertificate_openssl.cpp:236:29: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
     if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_RSA) {
                             ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
ssl/qsslcertificate_openssl.cpp:240:36: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
     } else if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_DSA) {
                                    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
ssl/qsslcertificate_openssl.cpp:245:36: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
     } else if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_EC) {
                                    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
ssl/qsslcertificate_openssl.cpp:250:36: error: invalid use of incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
     } else if (q_EVP_PKEY_type(pkey->type) == EVP_PKEY_DH) {
                                    ^~
In file included from /usr/include/openssl/crypto.h:31:0,
                 from /usr/include/openssl/comp.h:16,
                 from /usr/include/openssl/ssl.h:47,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/../../../../../src/network/ssl/qsslcontext_openssl_p.h:53,
                 from ../../include/QtNetwork/5.5.1/QtNetwork/private/qsslcontext_openssl_p.h:1,
                 from ssl/qsslsocket_p.h:55,
                 from ssl/qsslsocket_openssl_p.h:63,
                 from ssl/qsslsocket_openssl_symbols_p.h:64,
                 from ssl/qsslcertificate_openssl.cpp:35:
/usr/include/openssl/ossl_typ.h:93:16: note: forward declaration of ‘EVP_PKEY {aka struct evp_pkey_st}’
 typedef struct evp_pkey_st EVP_PKEY;
                ^~~~~~~~~~~
Makefile:21923: set di istruzioni per l'obiettivo ".obj/qsslcertificate_openssl.o" non riuscito
make[6]: *** [.obj/qsslcertificate_openssl.o] Errore 1
Makefile:191: set di istruzioni per l'obiettivo "sub-network-make_first" non riuscito
make[5]: *** [sub-network-make_first] Errore 2
Makefile:45: set di istruzioni per l'obiettivo "sub-src-make_first" non riuscito
make[4]: *** [sub-src-make_first] Errore 2
Makefile:50: set di istruzioni per l'obiettivo "module-qtbase-make_first" non riuscito
make[3]: *** [module-qtbase-make_first] Errore 2
CMakeFiles/Qt.dir/build.make:113: set di istruzioni per l'obiettivo "Qt-prefix/src/Qt-stamp/Qt-build" non riuscito
make[2]: *** [Qt-prefix/src/Qt-stamp/Qt-build] Errore 2
CMakeFiles/Makefile2:773: set di istruzioni per l'obiettivo "CMakeFiles/Qt.dir/all" non riuscito
make[1]: *** [CMakeFiles/Qt.dir/all] Errore 2
Makefile:105: set di istruzioni per l'obiettivo "all" non riuscito
make: *** [all] Errore 2
```

**Comment by h3r0n** [comment]
The culprit was some incompatibility with openssl 1.1.0 (that's the version shipped with Debian Stretch)
I solved this by installing `libssl-dev` 1.0.1 from Debian Jessie's repos.

there was anoter problem with make package looking for a file `/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.20` which isn't present on my system.
So used this as workaround
```
sudo cp /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.22 /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.20
```

So I obtained a deb package. When I installed it dpkg complained about the version number not beginning with a digit, so i edited this line in `build/CPackConfig.cmake`
```
SET(CPACK_DEBIAN_PACKAGE_VERSION "2.4-master")
```

So I managed to install the deb package.

**Comment by h3r0n** [comment]
But now the problem is that Cura fails to launch

```
~$ /usr/bin/cura
Traceback (most recent call last):
  File "/opt/cura/bin/cura", line 10, in <module>
    from UM.Platform import Platform
  File "/opt/cura/lib/python3/dist-packages/UM/__init__.py", line 9, in <module>
    from .Application import Application
  File "/opt/cura/lib/python3/dist-packages/UM/Application.py", line 9, in <module>
    from UM.Controller import Controller
  File "/opt/cura/lib/python3/dist-packages/UM/Controller.py", line 4, in <module>
    from UM.Scene.Scene import Scene
  File "/opt/cura/lib/python3/dist-packages/UM/Scene/Scene.py", line 4, in <module>
    from UM.Scene.SceneNode import SceneNode
  File "/opt/cura/lib/python3/dist-packages/UM/Scene/SceneNode.py", line 10, in <module>
    from UM.Mesh.MeshBuilder import MeshBuilder
  File "/opt/cura/lib/python3/dist-packages/UM/Mesh/MeshBuilder.py", line 4, in <module>
    from UM.Mesh.MeshData import MeshData
  File "/opt/cura/lib/python3/dist-packages/UM/Mesh/MeshData.py", line 13, in <module>
    import scipy.spatial
  File "/opt/cura/lib/python3/dist-packages/scipy/spatial/__init__.py", line 91, in <module>
    from .ckdtree import *
  File "__init__.pxd", line 155, in init ckdtree (scipy/spatial/ckdtree/ckdtree.cxx:22466)
ValueError: numpy.dtype has the wrong size, try recompiling
```

**Comment by Ghostkeeper** [comment]
Your Numpy (Python package) probably is too old. We use Numpy version 1.9.2 in our builds. What version are you using?

You can find your Numpy version via the command line this way:

    python3 -c "import numpy; print(numpy.__version__)"

Numpy binaries are backwards compatible, but not forwards compatible, so if you have something older than 1.9.2 it might not be compatible with what we build.

**Comment by h3r0n** [comment]
well, actually it's newer

```
~$ python3 -c "import numpy; print(numpy.__version__)"
1.12.0b1
```

**Comment by h3r0n** [comment]
> Compiling from source is the only option to install Cura 2.x on my distro, as the official precompiled binaries don't work too. So I'm stuck with Cura 15.04.6.

I was wrong: looks like there are universal appimages for both [Cura 2.3.x](https://software.ultimaker.com/?show=) and [Cura 2.4.0](http://software.ultimaker.com/Cura_open_beta/). Why are they so hidden?

Anyway let me now if you need more testing.

**Comment by Ghostkeeper** [comment]
The AppImages are new for Cura 2.4.0. They'll be published as the default Linux release once 2.4 is ready.

**Comment by LipuFei** [comment]
Closing this issue due to inactivity.
